shared/utility/skill/MotionGeneration.hpp

```diff
@@ -0,0 +1,442 @@
+            Eigen::Matrix<Scalar, 3, 1> step_placement;
+            step_placement.x() =
+                std::max(std::min(velocity_target.x() * step_period, step_limits.x()), -step_limits.x());
+            step_placement.y() =
+                std::max(std::min(velocity_target.y() * step_period, step_limits.y()), -step_limits.y());
+            step_placement.z() =
+                std::max(std::min(velocity_target.z() * step_period, step_limits.z()), -step_limits.z());
+
+            // X position trajectory
+            swingfoot_trajectory.add_waypoint(X, 0, Hps_start.translation().x(), 0);
+            swingfoot_trajectory.add_waypoint(X, step_apex_ratio * step_period, 0, velocity_target.x());
+            swingfoot_trajectory.add_waypoint(X, step_period, step_placement.x(), 0);
+
+            // Y position trajectory
+            swingfoot_trajectory.add_waypoint(Y, 0, Hps_start.translation().y(), 0);
+            swingfoot_trajectory.add_waypoint(Y,
+                                              step_apex_ratio * step_period,
+                                              get_foot_width_offset(),
+                                              velocity_target.y());
+            swingfoot_trajectory.add_waypoint(Y, step_period, get_foot_width_offset() + step_placement.y(), 0);
+
+            // Z position trajectory
+            swingfoot_trajectory.add_waypoint(Z, 0, Hps_start.translation().z(), 0);
+            swingfoot_trajectory.add_waypoint(Z, step_apex_ratio * step_period, step_height, 0);
+            swingfoot_trajectory.add_waypoint(Z, step_period, 0, 0);
+
+            // Roll trajectory
+            swingfoot_trajectory.add_waypoint(ROLL, 0, 0, 0);
+            swingfoot_trajectory.add_waypoint(ROLL, step_apex_ratio * step_period, 0, 0);
+            swingfoot_trajectory.add_waypoint(ROLL, step_period, 0, 0);
+
+            // Pitch trajectory
+            swingfoot_trajectory.add_waypoint(PITCH, 0, 0, 0);
+            swingfoot_trajectory.add_waypoint(PITCH, step_apex_ratio * step_period, 0, 0);
+            swingfoot_trajectory.add_waypoint(PITCH, step_period, 0, 0);
+
+            // Yaw trajectory
+            swingfoot_trajectory.add_waypoint(YAW, 0, 0, 0);
+            swingfoot_trajectory.add_waypoint(YAW,
+                                              step_apex_ratio * step_period,
+                                              velocity_target.z() * step_apex_ratio * step_period,
+                                              0);
+            swingfoot_trajectory.add_waypoint(YAW, step_period, step_placement.z(), 0);
+        }
+
+        /**
+         * @brief Generate torso trajectory.
+         * @param velocity_target Requested velocity target (dx, dy, dtheta).
+         * @param Hps_end Next foot placement.
+         * @return Trajectory of torso to follow to reach next torso placement.
+         */
+        void generate_torso_trajectory(const Eigen::Matrix<Scalar, 3, 1>& velocity_target) {
```

Comment:
these look like internal functions, if they're not used outside then put them in private